conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/TezosGovernanceOperations.scala

```diff
@@ -0,0 +1,203 @@
+import cats.implicits._
+
+/** Process blocks and voting data to compute details for
+  * the governance-related cycles
+  */
+object TezosGovernanceOperations extends LazyLogging {
+
+  /** Collects all relevant data for the governance
+    * voting process
+    *
+    * @param hash identifies a specific block
+    * @param metadata the block metadata
+    * @param proposalId a specific proposal under evaluation
+    * @param allRolls ballot rolls up to this block for the proposal
+    * @param singleBlockRolls ballot rolls included in this specific block
+    * @param ballotsPerLevel how many ballots for the single block level
+    * @param ballotsPerCycle how many baloots for a whole cycle
+    */
+  case class GovernanceAggregate(
+      hash: BlockHash,
+      metadata: BlockHeaderMetadata,
+      proposalId: Option[ProtocolId],
+      allRolls: VoteRollsCounts,
+      rollsPerLevel: VoteRollsCounts,
+      ballotsPerCycle: Option[Voting.BallotCounts],
+      ballotsPerLevel: Option[Voting.BallotCounts]
+  )
+
+  //wrappings to make the distinction between two different counts
+  private case class BallotsPerCycle(counts: Voting.BallotCounts) extends AnyVal
+  private case class BallotsPerLevel(counts: Voting.BallotCounts) extends AnyVal
+
+  /** Collects rolls of votes counted during ballots
+    *
+    * @param yay how many yays
+    * @param nay how many nays
+    * @param pass how many passes
+    */
+  case class VoteRollsCounts(yay: Int, nay: Int, pass: Int)
+
+  /** We take basic governance data and the blocks storing that
+    * with the goal of deriving more detailed aggregate data on
+    * the voting process.
+    * The computation needs to access the database to fetch essential
+    * ballot data
+    *
+    * SAFETY NOTE: make sure that the blocks passed to this method have
+    *              have a proper metadata value, different from the genesis one
+    *
+    * @param db needed to fetch previously stored voting data
+    * @param proposalsBlocks blocks we want to have governance data for, associated to the protocol of the proposal
+    * @param proposalsBallots votes cast on the blocks of interest,
+    *                         the keys should be a superset of the `proposalsBlocks` keys,
+    *                         so that the value is available for each of those blocks.
+    * @param levelsRolls listings of baker rolls involved, indexed per level, we expect all levels for the
+    *                         `proposalsBlocks` to be available, plus the level immediately before this batch.
+    *                          ideally this would be a `Level => List[Voting.BakerRolls]` (i.e. a total function)
+    * @param ec needed to compose concurrent operations
+    * @return aggregated data
+    */
+  def aggregateData(db: Database)(
+      proposalsBlocks: Map[Block, ProtocolId],
+      proposalsBallots: Map[Block, List[Voting.Ballot]],
+      levelsRolls: Map[Int, List[Voting.BakerRolls]]
```

Comment:
I had the same thoughts. I originally left it like this to avoid another dependency for the governance operator, both on database and on the node operator.

But it's probably cleaner like you suggest.